include/storage/fbw.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for flash buffered writes
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_FBW_H_
+#define ZEPHYR_INCLUDE_STORAGE_FBW_H_
+
+/**
+ * @brief Abstraction over flash buffered writes.
+ *
+ * @defgroup fbw flash buffered write Interface
+ * @{
+ */
+
+#include <stdbool.h>
+#include <storage/flash_map.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct fbw_ctx {
+	u8_t *buf;
+	size_t buf_len;
+	const struct flash_area *flash_area;
+	size_t bytes_written;
+	u16_t buf_bytes;
+#ifdef CONFIG_IMG_ERASE_PROGRESSIVELY
+	off_t off_last;
+#endif
+};
+
+/**
+ * @brief Initialize context needed for buffered write to flash.
+ *
+ * @param ctx context to be initialized
+ * @param area_id Area ID to write to
+ * @param offset Offset within area to write to
+ *
+ * @return  0 on success, negative errno code on fail
+ */
+int fbw_init(struct fbw_ctx *ctx, u8_t *buf, size_t buf_len, u8_t area_id,
+	     size_t offset);
+
+/**
+ * @brief Read number of bytes written to the flash.
+ *
+ * @param ctx context
+ *
+ * @return Number of bytes written to the image flash.
+ */
+size_t fbw_bytes_written(struct fbw_ctx *ctx);
+
+/**
+ * @brief  Process input buffers to be written to the image slot 1. flash
```

Comment:
These need a bit of a cleanup to remove flash_img specific language, right?

Reply:
Correct! Sorry for the sloppy work, I just threw this out so I could get your review before office time tomorrow :)

Reply:
Not sloppy at all! I really appreciate the handoff -- I know you're posting this as a draft.